File: 474-ones-and-zeroes/474-ones-and-zeroes.java
class Solution {
    public int findMaxForm(String[] strs, int m, int n) {
        int r = strs.length;
        int[][] dp = new int[m+1][n+1];
        
        //count zero and one for 1st index.
        int zero = 0, one = 0;
        for(int i=0;i<strs[0].length();i++) {
            if(strs[0].charAt(i) == '0') zero++;
            else one++;
        }
        
        //base case.
        for(int i=zero;i<=m;i++)
            for(int j=one;j<=n;j++)
                dp[i][j] = 1;
        
        for(int i=1;i<r;i++) {
            int[][] temp = new int[m+1][n+1];
            zero = one = 0; //counting number of zeros and ones.
            for(int j=0;j<strs[i].length();j++) {
                if(strs[i].charAt(j) == '0') zero++;
                else one++;
            }
            
            for(int j=0;j<=m;j++) {
                for(int k=0;k<=n;k++) {
                    int noPick = dp[j][k];
                    int pick = Integer.MIN_VALUE;
                    if(j - zero >= 0 && k - one >= 0)
                        pick = dp[j-zero][k-one] + 1;
                    temp[j][k] = Math.max(pick, noPick);
                }
            }
            dp = temp;
        }
        return dp[m][n];
    }
}

File: 474-ones-and-zeroes/474-ones-and-zeroes.cpp
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        int size = strs.size();
        vector<vector<vector<int>>> dp(size + 1, vector<vector<int>>(m + 1, vector<int>(n + 1, -1)));
        
        return helper(size-1, strs, n, m, dp);
    }
    
    int helper(int r, vector<string>& strs, int n, int m, vector<vector<vector<int>>>& dp) {
        int zeros = 0, ones = 0;
        
        for(auto el : strs[r]) { //counting one's and zero's in each string.
            if(el == '0') zeros++;
            else ones++;
        }
        
        if(r == 0) {
            if(m - zeros >= 0 && n - ones >= 0) return 1;
            return 0;
        }
        
        if(dp[r][m][n] != -1) return dp[r][m][n];
        
        int notPick = helper(r - 1, strs, n, m, dp);
        int pick = INT_MIN;
        
        if(m - zeros >= 0 && n - ones >= 0)
            pick = helper(r-1, strs, n - ones, m - zeros, dp) + 1;
        
        return dp[r][m][n] = max(pick, notPick);
    }
};